File: atcoder/other/tmu002/c.cpp
// C - ナップサック問題
// template
#include <bits/stdc++.h>
using namespace std;
typedef long long int ll;
typedef pair<long long, long long> LP;
#define INF 999999999
#define MOD 1000000007
#define REP(i, n) for(ll i = 0, i##_len = (n); i < i##_len; ++i)
#define REP_AB(i, a, b) for (ll i = ll(a); i <= ll(b); ++i)
#define ALL(v) v.begin(), v.end()
#define SORT(v) sort(ALL(v))
#define UNIQUE(v) sort(ALL(v));v.erase(unique(ALL(v)), v.end());
#define SIZE(x) ((ll)(x).size())
#define REVERSE(v) reverse(ALL(v))
// true / false
#define BIN_SEARCH(v, a) binary_search(ALL(v), a)
// index
#define BIN_LEFT(v, a) (lower_bound(ALL(v), a) - v.begin())
#define BIN_RIGHT(v, a) (upper_bound(ALL(v), a) - v.begin())
#define BIN_INSERT(v, a) (v.insert(v.begin() + BIN_LEFT(v, a), a))
#define DEL(v, i) v.erase(v.begin() + i)
#define INSERT(v, i, a) v.insert(v.begin() + i, a)
ll gcd(ll a,ll b){return b?gcd(b,a%b):a;}
ll lcm(ll a,ll b){return (a / gcd(a, b)) * b;}
int dx[4]={1,0,-1,0};
int dy[4]={0,1,0,-1};
#define out cout
#define in cin
template<class T>bool chmax(T &a, const T &b) { if (a<b) { a=b; return 1; } return 0; }
template<class T>bool chmin(T &a, const T &b) { if (b<a) { a=b; return 1; } return 0; }
// template end

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(0);

    ll n, W;
    in >> n >> W;
    ll w[n+1] = {};
    ll v[n+1] = {};

    ll max_w = 0;
    ll max_v = 0;
    REP(i, n){
        in >> v[i + 1]  >> w[i + 1];
        chmax(max_w, w[i+1]);
        chmax(max_v, v[i+1]);
    }

    if(max_w <= 1000){
        ll dp[n + 1][W + 1] = {};

        REP_AB(i, 1, n){
            REP_AB(j, 0, W){
                dp[i][j] = dp[i-1][j];
                if(j - 1 < 0) continue;
                dp[i][j] = max(dp[i][j], dp[i][j-1]);
                if(j-w[i] < 0) continue;
                dp[i][j] = max(dp[i][j], dp[i-1][j-w[i]] + v[i]);
            }
        }
        out << dp[n][W] << endl;
    }else if(max_v <= 1000){
        ll sum_v = 0;
        REP(i, n+1){
            sum_v += v[i];
        }
        ll dp[n+1][sum_v+1] = {};
        REP(i, n+1){
            REP(j, sum_v+1){
                dp[i][j] = INF;
            }
            dp[i][0] = 0;
        }
        REP_AB(i, 1, n){
            REP(j, v[i]) dp[i][j] = dp[i-1][j];
            REP_AB(j, v[i], sum_v){
                // dp[n番目][価値jの時] = 価値をjにできる最小のw
                dp[i][j] = dp[i-1][j];
                if(dp[i-1][j-v[i]] > W) continue;
                dp[i][j] = min(dp[i-1][j-v[i]] + w[i], dp[i][j]);
            }
        }
        ll ans = 0;
        REP(i, sum_v+1){
            if(dp[n][i] <= W){
                ans = i;
            }
        }
        out << ans << endl;
    }else{
        // 半分全列挙
        // ll a = int(n/2);
        // ll b = n - a;

        // REP(i, n+1){
        //     if(i < a){
        //         REP_AB(j, )
        //     }else{

        //     }
        // }
    }
}

File: atcoder/other/tmu002/c.py
import bisect

INF = 10 ** 15
n, W = map(int, input().split())
w = []
v = []
max_w = 0
max_v = 0
for i in range(n):
    v_, w_ = map(int, input().split())
    if max_v < v_:
        max_v = v_
    if max_w < w_:
        max_w = w_
    w.append(w_)
    v.append(v_)

# N <= 30
def solve1():
    global n, W, items
    # 半分全列挙
    a = n // 2
    b = n - a

    items1 = []
    items2 = []
    for i in range(2 ** a):
        w_ = 0
        v_ = 0
        for j in range(a):
            if i & 2 ** j > 0:
                w_ += w[j]
                v_ += v[j]
        items1.append([w_, v_])
    items1 = sorted(items1, key=lambda x: (x[0], -x[1]))
    for i in range(2 ** b):
        w_ = 0
        v_ = 0
        for j in range(b):
            if i & 2 ** j > 0:
                w_ += w[a + j]
                v_ += v[a + j]
        items2.append([w_, v_])
    items2 = sorted(items2, key=lambda x: (x[0], -x[1]))

    a_items = []
    pre_v = 0
    for i in range(len(items1)):
        if items1[i][0] > W:
            break
        if items1[i][1] > pre_v:
            a_items.append(items1[i])
            pre_v = items1[i][1]
    a_items.insert(0, [0, 0])

    b_items = []
    pre_v = 0
    for i in range(len(items2)):
        if items2[i][0] > W:
            break
        if items2[i][1] > pre_v:
            b_items.append(items2[i])
            pre_v = items2[i][1]
    b_items.insert(0, [0, 0])
    b_items.append([10 ** 20, 0])
    b_w = [x[0] for x in b_items]

    ans = 0
    for i in range(len(a_items)):
        w_ = W - a_items[i][0]
        idx = bisect.bisect_left(b_w, w_) - 1
        if a_items[i][0] + b_items[idx][0] <= W:
            ans = max(ans, a_items[i][1] + b_items[idx][1])
        idx = bisect.bisect_left(b_w, w_)
        if a_items[i][0] + b_items[idx][0] <= W:
            ans = max(ans, a_items[i][1] + b_items[idx][1])
    print(ans)


# max_w <= 1000
def solve2():
    global n, W, items
    # wのdp
    dp = [[0 for _ in range(W + 1)] for _ in range(n + 1)]

    for i in range(n):
        for j in range(w[i]):
            dp[i + 1][j] = dp[i][j]
        for j in range(w[i], W + 1):
            dp[i + 1][j] = max(dp[i][j], dp[i + 1][j - 1], dp[i][j - w[i]] + v[i])
    ans = 0
    for i in range(W + 1):
        ans = max(ans, dp[n][i])
    print(ans)


# max_v <= 1000
def solve3():
    global n, W, items
    # vのdp
    sum_v = 0
    for i in range(n):
        sum_v += v[i]
    dp = [[INF for _ in range(sum_v + 1)] for _ in range(n + 1)]

    for i in range(n + 1):
        dp[i][0] = 0
    for i in range(n):
        for j in range(v[i]):
            dp[i + 1][j] = dp[i][j]
        for j in range(v[i], sum_v + 1):
            dp[i + 1][j] = min(dp[i][j], dp[i][j - v[i]] + w[i])
    ans = 0
    for i in range(len(dp[n])):
        if dp[n][i] <= W:
            ans = i
    print(ans)


if n <= 30:
    solve1()
elif max_w <= 1000:
    solve2()
else:
    solve3()
